Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp

```diff
@@ -0,0 +1,701 @@
+      break;
+    }
+
+    // set default size to 0 to avoid issue
+    datasize = 0;
+    const bool can_go = (GetTickCountStd() - ref->poke_time > 100 || ref->window_size > 2000);
+    std::array<u8, 2048> buffer;
+    if (tcp_buffer != nullptr && ref->ready && can_go)
+      st = ref->tcp_socket.receive(&buffer[0], MAX_TCP_LENGTH, datasize);
+
+    if (datasize > 0)
+    {
+      Common::TCPPacket packet = Common::TCPPacket(ref->bba_mac, ref->my_mac);
+      packet.ip_header = Common::IPv4Header((u16)(datasize + 20), IPPROTO_TCP, ref->from, ref->to);
+      packet.tcp_header =
+          Common::TCPHeader(ref->from, ref->to, ref->seq_num, ref->ack_num, TCP_FLAG_ACK);
+      packet.data = std::vector<u8>(buffer.begin(), buffer.begin() + datasize);
+
+      // build buffer
+      tcp_buffer->seq_id = ref->seq_num;
+      tcp_buffer->tick = GetTickCountStd();
+      tcp_buffer->data = packet.Build();
+      tcp_buffer->seq_id = ref->seq_num;
+      tcp_buffer->used = true;
+      ref->seq_num += (u32)datasize;
+      ref->poke_time = GetTickCountStd();
+      return tcp_buffer->data;
+    }
+    if (GetTickCountStd() - ref->delay > 3000)
+    {
+      if (st == sf::Socket::Disconnected || st == sf::Socket::Error)
+      {
+        ref->ip = 0;
+        ref->tcp_socket.disconnect();
+        return BuildFINFrame(ref);
+      }
+    }
+    break;
+  }
+
+  return std::nullopt;
+}
+
+void CEXIETHERNET::BuiltInBBAInterface::ReadThreadHandler(CEXIETHERNET::BuiltInBBAInterface* self)
+{
+  while (!self->m_read_thread_shutdown.IsSet())
+  {
+    // make thread less cpu hungry
+    std::this_thread::sleep_for(std::chrono::milliseconds(1));
+
+    if (!self->m_read_enabled.IsSet())
+      continue;
+    size_t datasize = 0;
+
+    u8 wp = self->m_eth_ref->page_ptr(BBA_RWP);
+    const u8 rp = self->m_eth_ref->page_ptr(BBA_RRP);
+    if (rp > wp)
+      wp += 16;
+
+    if ((wp - rp) >= 8)
+      continue;
+
+    std::lock_guard<std::mutex> lock(self->m_mtx);
+    // process queue file first
+    if (self->m_queue_read != self->m_queue_write)
+    {
+      datasize = self->m_queue_data[self->m_queue_read].size();
+      if (datasize > BBA_RECV_SIZE)
+      {
+        ERROR_LOG_FMT(SP1, "Frame size is exceiding BBA capacity, frame stack might be corrupted"
+                           "Killing Dolphin...");
+        std::exit(0);
+      }
+      std::memcpy(self->m_eth_ref->mRecvBuffer.get(), self->m_queue_data[self->m_queue_read].data(),
+                  datasize);
+      self->m_queue_read++;
+      self->m_queue_read &= 15;
+    }
+    else
+    {
+      // test connections data
+      for (auto& net_ref : self->network_ref)
+      {
+        if (net_ref.ip == 0)
+          continue;
+        const auto socket_data = TryGetDataFromSocket(&net_ref);
+        if (socket_data.has_value())
+        {
+          datasize = socket_data.value().size();
+          std::memcpy(self->m_eth_ref->mRecvBuffer.get(), socket_data.value().data(), datasize);
```

Comment:
tryed, but VS is piss off

Reply:
'data' n'est pas membre de 'std::optional<std::vector<u8,std::allocator<sf::Uint8>>>'

in english: data is not a member of std.....

Reply:
Make sure you're using `socket_data->data()` instead of `socket_data.data()`. The former gets the vector contained inside the optional and then calls vector's data() function (which is what you want), while the latter tries to call optional's non-existent data() function.